# Завдання 1

In [ ]:
%%bash
apt-get -qq update
apt-get -qq install -y postgresql postgresql-contrib > /dev/null

service postgresql start

sudo -u postgres psql -c "ALTER USER postgres PASSWORD 'postgres';"

sudo -u postgres psql -c "DROP DATABASE IF EXISTS hw4_db;"
sudo -u postgres psql -c "CREATE DATABASE hw4_db;"

In [ ]:
%pip install -q psycopg2-binary sqlalchemy pandas kagglehub

Note: you may need to restart the kernel to use updated packages.


In [ ]:
import os

import pandas as pd
from sqlalchemy import create_engine, text

DB_URL = "postgresql+psycopg2://postgres:postgres@localhost:5432/hw4_db"
engine = create_engine(DB_URL)

with engine.connect() as conn:
    print(conn.execute(text('SELECT version()')).scalar_one())

PostgreSQL 16.2 on aarch64-apple-darwin23.5.0, compiled by Apple clang version 15.0.0 (clang-1500.0.40.1), 64-bit


In [ ]:
import kagglehub

dataset_dir = kagglehub.dataset_download('olistbr/brazilian-ecommerce')
print(dataset_dir)
print(sorted(os.listdir(dataset_dir)))

/Users/vladyslavsvietlov/.venvs/goit-rdb-hw-04-py312/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


/Users/vladyslavsvietlov/.cache/kagglehub/datasets/olistbr/brazilian-ecommerce/versions/2
['olist_customers_dataset.csv', 'olist_geolocation_dataset.csv', 'olist_order_items_dataset.csv', 'olist_order_payments_dataset.csv', 'olist_order_reviews_dataset.csv', 'olist_orders_dataset.csv', 'olist_products_dataset.csv', 'olist_sellers_dataset.csv', 'product_category_name_translation.csv']


In [ ]:
raw_sources = {
    'olist_customers_raw': 'olist_customers_dataset.csv',
    'olist_orders_raw': 'olist_orders_dataset.csv',
    'olist_order_items_raw': 'olist_order_items_dataset.csv',
    'olist_products_raw': 'olist_products_dataset.csv',
    'olist_sellers_raw': 'olist_sellers_dataset.csv',
    'olist_order_reviews_raw': 'olist_order_reviews_dataset.csv',
}

raw_counts = {}
for table_name, csv_name in raw_sources.items():
    df = pd.read_csv(os.path.join(dataset_dir, csv_name))
    df.to_sql(table_name, engine, if_exists='replace', index=False, chunksize=10_000)
    raw_counts[table_name] = len(df)

pd.Series(raw_counts, name='n_rows').to_frame()

,n_rows
olist_customers_raw,99441
olist_orders_raw,99441
olist_order_items_raw,112650
olist_products_raw,32951
olist_sellers_raw,3095
olist_order_reviews_raw,99224


In [ ]:
ddl = """
DROP TABLE IF EXISTS customer_segments CASCADE;
DROP TABLE IF EXISTS seller_score CASCADE;
DROP TABLE IF EXISTS seller_alerts CASCADE;
DROP TABLE IF EXISTS olist_customer_dim CASCADE;
DROP TABLE IF EXISTS olist_order_reviews CASCADE;
DROP TABLE IF EXISTS olist_order_items CASCADE;
DROP TABLE IF EXISTS olist_orders CASCADE;
DROP TABLE IF EXISTS olist_customers CASCADE;
DROP TABLE IF EXISTS olist_products CASCADE;
DROP TABLE IF EXISTS olist_sellers CASCADE;

CREATE TABLE olist_customers (
    customer_id              TEXT PRIMARY KEY,
    customer_unique_id       TEXT NOT NULL,
    customer_zip_code_prefix INTEGER,
    customer_city            TEXT,
    customer_state           CHAR(2) NOT NULL
);

CREATE INDEX idx_olist_customers_unique_id
    ON olist_customers (customer_unique_id);

CREATE TABLE olist_orders (
    order_id                      TEXT PRIMARY KEY,
    customer_id                   TEXT NOT NULL REFERENCES olist_customers(customer_id),
    order_status                  TEXT NOT NULL,
    order_purchase_timestamp      TIMESTAMP NOT NULL,
    order_approved_at             TIMESTAMP,
    order_delivered_carrier_date  TIMESTAMP,
    order_delivered_customer_date TIMESTAMP,
    order_estimated_delivery_date TIMESTAMP,
    CHECK (order_status IN (
        'created', 'approved', 'invoiced', 'processing',
        'shipped', 'delivered', 'unavailable', 'canceled'
    ))
);

CREATE TABLE olist_products (
    product_id                 TEXT PRIMARY KEY,
    product_category_name      TEXT,
    product_name_length        INTEGER CHECK (product_name_length IS NULL OR product_name_length >= 0),
    product_description_length INTEGER CHECK (product_description_length IS NULL OR product_description_length >= 0),
    product_photos_qty         INTEGER CHECK (product_photos_qty IS NULL OR product_photos_qty >= 0),
    product_weight_g           INTEGER CHECK (product_weight_g IS NULL OR product_weight_g >= 0),
    product_length_cm          INTEGER CHECK (product_length_cm IS NULL OR product_length_cm >= 0),
    product_height_cm          INTEGER CHECK (product_height_cm IS NULL OR product_height_cm >= 0),
    product_width_cm           INTEGER CHECK (product_width_cm IS NULL OR product_width_cm >= 0)
);

CREATE TABLE olist_sellers (
    seller_id              TEXT PRIMARY KEY,
    seller_zip_code_prefix INTEGER,
    seller_city            TEXT,
    seller_state           CHAR(2) NOT NULL
);

CREATE TABLE olist_order_items (
    order_id            TEXT NOT NULL REFERENCES olist_orders(order_id),
    order_item_id       INTEGER NOT NULL,
    product_id          TEXT REFERENCES olist_products(product_id),
    seller_id           TEXT REFERENCES olist_sellers(seller_id),
    shipping_limit_date TIMESTAMP,
    price               NUMERIC(12, 2) CHECK (price IS NULL OR price >= 0),
    freight_value       NUMERIC(12, 2) CHECK (freight_value IS NULL OR freight_value >= 0),
    PRIMARY KEY (order_id, order_item_id)
);

CREATE INDEX idx_olist_order_items_seller_order
    ON olist_order_items (seller_id, order_id);

CREATE TABLE olist_order_reviews (
    review_row_id            BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
    review_id                TEXT,
    order_id                 TEXT NOT NULL REFERENCES olist_orders(order_id),
    review_score             SMALLINT CHECK (review_score BETWEEN 1 AND 5),
    review_comment_title     TEXT,
    review_comment_message   TEXT,
    review_creation_date     TIMESTAMP,
    review_answer_timestamp  TIMESTAMP
);

CREATE INDEX idx_olist_order_reviews_order_id
    ON olist_order_reviews (order_id);
"""

with engine.begin() as conn:
    conn.exec_driver_sql(ddl)

Типізована схема з шести таблиць: PK, FK, `CHECK`-обмеження та індекси під типові з'єднання. `DROP TABLE IF EXISTS ... CASCADE` (включно з таблицями наступних завдань) робить клітинку ідемпотентною. У `olist_order_reviews` штучний identity-ключ `review_row_id`, бо `review_id` не унікальний.

In [ ]:
transfer_sql = """
INSERT INTO olist_customers
SELECT
    customer_id,
    customer_unique_id,
    customer_zip_code_prefix::INTEGER,
    customer_city,
    customer_state::CHAR(2)
FROM olist_customers_raw;

INSERT INTO olist_orders
SELECT
    order_id,
    customer_id,
    order_status,
    order_purchase_timestamp::TIMESTAMP,
    order_approved_at::TIMESTAMP,
    order_delivered_carrier_date::TIMESTAMP,
    order_delivered_customer_date::TIMESTAMP,
    order_estimated_delivery_date::TIMESTAMP
FROM olist_orders_raw;

INSERT INTO olist_products
SELECT
    product_id,
    product_category_name,
    product_name_lenght::INTEGER,
    product_description_lenght::INTEGER,
    product_photos_qty::INTEGER,
    product_weight_g::INTEGER,
    product_length_cm::INTEGER,
    product_height_cm::INTEGER,
    product_width_cm::INTEGER
FROM olist_products_raw;

INSERT INTO olist_sellers
SELECT
    seller_id,
    seller_zip_code_prefix::INTEGER,
    seller_city,
    seller_state::CHAR(2)
FROM olist_sellers_raw;

INSERT INTO olist_order_items
SELECT
    order_id,
    order_item_id::INTEGER,
    product_id,
    seller_id,
    shipping_limit_date::TIMESTAMP,
    price::NUMERIC(12, 2),
    freight_value::NUMERIC(12, 2)
FROM olist_order_items_raw;

INSERT INTO olist_order_reviews (
    review_id,
    order_id,
    review_score,
    review_comment_title,
    review_comment_message,
    review_creation_date,
    review_answer_timestamp
)
SELECT
    review_id,
    order_id,
    review_score::SMALLINT,
    review_comment_title,
    review_comment_message,
    review_creation_date::TIMESTAMP,
    review_answer_timestamp::TIMESTAMP
FROM olist_order_reviews_raw;
"""

with engine.begin() as conn:
    conn.exec_driver_sql(transfer_sql)

Дані перенесено з raw у типізовані таблиці через `INSERT ... SELECT` з явним приведенням типів, у порядку «батьки → діти» для коректної перевірки FK. Помилкові назви колонок CSV (`product_name_lenght`, `product_description_lenght`) нормалізовано до `*_length`. Усе виконується в одній транзакції; для повторного запуску спершу перезапустити DDL.

In [ ]:
smoke_test = pd.read_sql(text("""
SELECT 'customers' AS table_name, COUNT(*) AS n_rows FROM olist_customers
UNION ALL SELECT 'orders', COUNT(*) FROM olist_orders
UNION ALL SELECT 'order_items', COUNT(*) FROM olist_order_items
UNION ALL SELECT 'products', COUNT(*) FROM olist_products
UNION ALL SELECT 'sellers', COUNT(*) FROM olist_sellers
UNION ALL SELECT 'reviews', COUNT(*) FROM olist_order_reviews
ORDER BY table_name
"""), engine)

raw_by_name = {
    'customers': raw_counts['olist_customers_raw'],
    'orders': raw_counts['olist_orders_raw'],
    'order_items': raw_counts['olist_order_items_raw'],
    'products': raw_counts['olist_products_raw'],
    'sellers': raw_counts['olist_sellers_raw'],
    'reviews': raw_counts['olist_order_reviews_raw'],
}
smoke_test['raw_rows'] = smoke_test['table_name'].map(raw_by_name)
smoke_test['diff'] = smoke_test['raw_rows'] - smoke_test['n_rows']
smoke_test

,table_name,n_rows,raw_rows,diff
0,customers,99441,99441,0
1,order_items,112650,112650,0
2,orders,99441,99441,0
3,products,32951,32951,0
4,reviews,99224,99224,0
5,sellers,3095,3095,0


Smoke-test: у всіх шести таблицях `diff` = 0, тобто typed і raw мають однакову кількість рядків і жоден рядок не відкинуто через FK, `CHECK` чи PK. Однакова кількість клієнтів і замовлень (99 441) — бо в Olist `customer_id` створюється на кожне замовлення.

# Завдання 2

In [ ]:
seller_alerts_ddl = """
DROP TABLE IF EXISTS seller_alerts CASCADE;

CREATE TABLE seller_alerts (
    alert_id   BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    seller_id  TEXT NOT NULL REFERENCES olist_sellers(seller_id),
    alert_type TEXT NOT NULL CHECK (alert_type IN ('late_delivery', 'low_rating', 'data_quality')),
    severity   SMALLINT NOT NULL CHECK (severity BETWEEN 1 AND 5),
    details    TEXT,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
)
"""

insert_alert_sql = """
WITH seller_orders AS (
    SELECT DISTINCT
        oi.seller_id,
        oi.order_id
    FROM olist_order_items AS oi
),
late_sellers AS (
    SELECT
        so.seller_id,
        COUNT(DISTINCT o.order_id) AS late_orders
    FROM seller_orders AS so
    JOIN olist_orders AS o
        ON o.order_id = so.order_id
    WHERE o.order_delivered_customer_date IS NOT NULL
      AND o.order_estimated_delivery_date IS NOT NULL
      AND o.order_delivered_customer_date::DATE > o.order_estimated_delivery_date::DATE
    GROUP BY so.seller_id
    ORDER BY late_orders DESC, so.seller_id
    LIMIT 1
)
INSERT INTO seller_alerts (seller_id, alert_type, severity, details)
SELECT
    seller_id,
    'late_delivery',
    4,
    'Seller has ' || late_orders || ' late delivered orders in the loaded dataset'
FROM late_sellers
RETURNING alert_id, seller_id, alert_type, severity, created_at
"""

with engine.begin() as conn:
    conn.exec_driver_sql(seller_alerts_ddl)
    result = conn.exec_driver_sql(insert_alert_sql)
    new_alerts = pd.DataFrame(result.fetchall(), columns=list(result.keys()))

new_alerts

,alert_id,seller_id,alert_type,severity,created_at
0,1,4a3ca9315b744ce9f8e9374361493884,late_delivery,4,2026-09-30 23:31:34.175769+03:00


`INSERT ... RETURNING`: у `seller_alerts` додано alert `late_delivery` для продавця з найбільшою кількістю запізнілих замовлень (пари `seller_id × order_id` попередньо дедублікуються). `RETURNING` одразу повертає створений рядок (`alert_id`, `created_at`) — готовий audit-log запис без окремого `SELECT`. `DROP TABLE IF EXISTS ... CASCADE` + `CREATE TABLE` роблять клітинку idempotent: після кожного перезапуску в таблиці рівно один alert.

In [ ]:
seller_score_ddl = """
CREATE TABLE IF NOT EXISTS seller_score (
    seller_id   TEXT PRIMARY KEY REFERENCES olist_sellers(seller_id),
    avg_rating  NUMERIC(3, 2),
    n_reviews   INTEGER NOT NULL DEFAULT 0 CHECK (n_reviews >= 0),
    tier        TEXT CHECK (tier IN ('gold', 'silver', 'bronze')),
    updated_at  TIMESTAMPTZ NOT NULL DEFAULT NOW()
)
"""

upsert_score_sql = """
WITH seller_orders AS (
    SELECT DISTINCT
        oi.seller_id,
        oi.order_id
    FROM olist_order_items AS oi
),
review_per_order AS (
    SELECT
        rv.order_id,
        AVG(rv.review_score)::NUMERIC(3, 2) AS order_review_score
    FROM olist_order_reviews AS rv
    WHERE rv.review_score IS NOT NULL
    GROUP BY rv.order_id
),
seller_rating AS (
    SELECT
        so.seller_id,
        AVG(rpo.order_review_score)::NUMERIC(3, 2) AS avg_rating,
        COUNT(rpo.order_id) AS n_reviews
    FROM seller_orders AS so
    LEFT JOIN review_per_order AS rpo
        ON rpo.order_id = so.order_id
    GROUP BY so.seller_id
)
INSERT INTO seller_score (seller_id, avg_rating, n_reviews, tier)
SELECT
    seller_id,
    avg_rating,
    n_reviews,
    CASE
        WHEN avg_rating IS NULL THEN NULL
        WHEN avg_rating >= 4.5 THEN 'gold'
        WHEN avg_rating >= 3.5 THEN 'silver'
        ELSE 'bronze'
    END AS tier
FROM seller_rating
ON CONFLICT (seller_id) DO UPDATE
SET avg_rating = EXCLUDED.avg_rating,
    n_reviews  = EXCLUDED.n_reviews,
    tier       = EXCLUDED.tier,
    updated_at = NOW()
RETURNING seller_id, avg_rating, n_reviews, tier, updated_at
"""

with engine.begin() as conn:
    conn.exec_driver_sql(seller_score_ddl)
    result = conn.exec_driver_sql(upsert_score_sql)
    upserted_scores = pd.DataFrame(result.fetchall(), columns=list(result.keys()))

print(len(upserted_scores))
print(upserted_scores['tier'].value_counts(dropna=False))
upserted_scores.head()

3095
tier
silver    1568
gold       983
bronze     539
NaN          5
Name: count, dtype: int64


,seller_id,avg_rating,n_reviews,tier,updated_at
0,a61cc04793308395a840807104365121,5.00,1,gold,2026-09-30 23:31:34.255670+03:00
1,6d2f2e3b539480db1e0842b3a4e32e6e,3.67,3,silver,2026-09-30 23:31:34.255670+03:00
2,c53bcd3be457a342a97e39e5a9f0be22,3.20,5,bronze,2026-09-30 23:31:34.255670+03:00
3,9591fc341b1bfb7ef561e2968ec6e011,1.00,1,bronze,2026-09-30 23:31:34.255670+03:00
4,f9ec7093df3a7b346b7bcf7864069ca3,4.78,9,gold,2026-09-30 23:31:34.255670+03:00


`INSERT ... ON CONFLICT DO UPDATE`: `seller_score` заповнюється середнім рейтингом, кількістю відгуків і тіром продавця.

1. **`SELECT DISTINCT` у `seller_orders`.** `olist_order_items` має рядок на позицію, тож пара `seller_id × order_id` може повторюватися; без дедуплікації один відгук врахувався б кілька разів.
2. **Агрегація відгуків до `order_id`.** Замовлення може мати кілька відгуків; `review_per_order` дає одну оцінку на замовлення, тож join не розмножує рядки.
3. **Ідемпотентність.** `ON CONFLICT (seller_id) DO UPDATE` не створює дублікатів, а перезаписує значення перерахованими з поточних даних.
4. **`NULL` / `0` / `NULL`.** Для продавців без відгуків `LEFT JOIN` дає `AVG` = `NULL`, `COUNT` = 0, а `CASE` — `NULL`-тір.

**Обмеження датасету:** відгук у Olist прив'язаний до замовлення, а не до продавця, тому відгук замовлення з кількома продавцями приписується кожному з них — прийнятне навчальне спрощення.

In [ ]:
is_late_sql = """
WITH delivery_flags AS (
    SELECT
        order_id,
        (
            order_delivered_customer_date IS NOT NULL
            AND order_estimated_delivery_date IS NOT NULL
            AND order_delivered_customer_date::DATE > order_estimated_delivery_date::DATE
        ) AS is_late_calc
    FROM olist_orders
)
UPDATE olist_orders AS o
SET is_late = f.is_late_calc
FROM delivery_flags AS f
WHERE f.order_id = o.order_id
RETURNING o.order_id, o.customer_id, o.is_late
"""

with engine.begin() as conn:
    conn.exec_driver_sql('ALTER TABLE olist_orders ADD COLUMN IF NOT EXISTS is_late BOOLEAN NOT NULL DEFAULT FALSE')
    result = conn.exec_driver_sql(is_late_sql)
    updated_orders = pd.DataFrame(result.fetchall(), columns=list(result.keys()))

print(len(updated_orders))
print(updated_orders['is_late'].value_counts())
updated_orders.head()

99441
is_late
False    92906
True      6535
Name: count, dtype: int64


,order_id,customer_id,is_late
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,False
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,False
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,False
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,False
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,False


`UPDATE ... FROM`: колонка `is_late` (`ADD COLUMN IF NOT EXISTS`) оновлюється з CTE `delivery_flags` — `TRUE`, якщо фактична доставка пізніша за очікувану. Ідемпотентно: прапорець щоразу перераховується з поточних даних і може стати як `TRUE`, так і `FALSE`. `RETURNING` показує оновлені замовлення.

In [ ]:
purge_alerts_sql = """
DELETE FROM seller_alerts
WHERE created_at < NOW() - INTERVAL '30 days'
RETURNING alert_id, seller_id, alert_type, created_at
"""

with engine.begin() as conn:
    result = conn.exec_driver_sql(purge_alerts_sql)
    purged_alerts = pd.DataFrame(result.fetchall(), columns=list(result.keys()))

print(len(purged_alerts))
purged_alerts

0


,alert_id,seller_id,alert_type,created_at


`DELETE ... RETURNING`: видаляє alert-и, старші за 30 днів; `RETURNING` повертає видалені рядки для audit-log. Порожній результат очікуваний — усі alert-и щойно створені. Перезапуск безпечний: видаляються лише застарілі рядки.

# Завдання 3

In [ ]:
customer_dim_ddl = """
CREATE TABLE IF NOT EXISTS olist_customer_dim (
    customer_unique_id TEXT PRIMARY KEY,
    latest_customer_id TEXT REFERENCES olist_customers(customer_id),
    customer_state     CHAR(2),
    customer_city      TEXT
)
"""

populate_customer_dim_sql = """
INSERT INTO olist_customer_dim (
    customer_unique_id,
    latest_customer_id,
    customer_state,
    customer_city
)
SELECT DISTINCT ON (c.customer_unique_id)
    c.customer_unique_id,
    c.customer_id AS latest_customer_id,
    c.customer_state,
    c.customer_city
FROM olist_customers AS c
LEFT JOIN olist_orders AS o
    ON o.customer_id = c.customer_id
ORDER BY
    c.customer_unique_id,
    o.order_purchase_timestamp DESC NULLS LAST,
    c.customer_id
ON CONFLICT (customer_unique_id) DO UPDATE
SET latest_customer_id = EXCLUDED.latest_customer_id,
    customer_state     = EXCLUDED.customer_state,
    customer_city      = EXCLUDED.customer_city
"""

with engine.begin() as conn:
    conn.exec_driver_sql(customer_dim_ddl)
    conn.exec_driver_sql(populate_customer_dim_sql)

pd.read_sql(text("""
SELECT
    (SELECT COUNT(*) FROM olist_customers) AS customers_rows,
    (SELECT COUNT(DISTINCT customer_unique_id) FROM olist_customers) AS unique_customers,
    (SELECT COUNT(*) FROM olist_customer_dim) AS customer_dim_rows
"""), engine)

,customers_rows,unique_customers,customer_dim_rows
0,99441,96096,96096


`olist_customer_dim` — один рядок на `customer_unique_id`; `DISTINCT ON` залишає дані з найновішого замовлення покупця (детермінований tie-break за `customer_id`). `CREATE TABLE IF NOT EXISTS` і `ON CONFLICT DO UPDATE` роблять клітинку перезапусканою.

In [ ]:
customer_segments_ddl = """
CREATE TABLE IF NOT EXISTS customer_segments (
    segment_id          BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,

    customer_unique_id  TEXT NOT NULL,
    segment_name        TEXT NOT NULL
        CHECK (segment_name IN ('VIP', 'regular', 'new', 'inactive')),

    rfm_score           SMALLINT NOT NULL
        CHECK (rfm_score BETWEEN 1 AND 5),

    recency_days        INTEGER
        CHECK (recency_days IS NULL OR recency_days >= 0),

    monetary_value      NUMERIC(12, 2) NOT NULL DEFAULT 0
        CHECK (monetary_value >= 0),

    n_orders            INTEGER NOT NULL DEFAULT 0
        CHECK (n_orders >= 0),

    assigned_at         TIMESTAMPTZ NOT NULL DEFAULT NOW(),

    UNIQUE (customer_unique_id)
)
"""

customer_segments_fk_sql = """
ALTER TABLE customer_segments
    ADD CONSTRAINT fk_customer_segments_customer_unique
    FOREIGN KEY (customer_unique_id)
    REFERENCES olist_customer_dim(customer_unique_id)
    ON DELETE CASCADE
"""

with engine.begin() as conn:
    conn.exec_driver_sql(customer_segments_ddl)
    conn.exec_driver_sql('ALTER TABLE customer_segments DROP CONSTRAINT IF EXISTS fk_customer_segments_customer_unique')
    conn.exec_driver_sql(customer_segments_fk_sql)

DDL `customer_segments`: identity PK `segment_id`, `UNIQUE (customer_unique_id)` (ціль для `ON CONFLICT`), `CHECK`-обмеження на сегмент, `rfm_score` 1–5 і невід'ємні значення, FK на `olist_customer_dim` з `ON DELETE CASCADE`. `DROP CONSTRAINT IF EXISTS` перед додаванням FK дозволяє перезапуск.

In [ ]:
segmentation_sql = """
WITH reference_date AS (
    SELECT (MAX(order_purchase_timestamp)::DATE + 1) AS as_of_date
    FROM olist_orders
),
order_totals AS (
    SELECT
        o.order_id,
        o.customer_id,
        o.order_purchase_timestamp::DATE AS order_date,
        COALESCE(SUM(oi.price + oi.freight_value), 0)::NUMERIC(12, 2) AS order_total
    FROM olist_orders AS o
    LEFT JOIN olist_order_items AS oi
        ON oi.order_id = o.order_id
    GROUP BY
        o.order_id,
        o.customer_id,
        o.order_purchase_timestamp
),
customer_stats AS (
    SELECT
        c.customer_unique_id,
        COUNT(DISTINCT ot.order_id) AS n_orders,
        COALESCE(SUM(ot.order_total), 0)::NUMERIC(12, 2) AS monetary_value,
        MAX(ot.order_date) AS last_order_date
    FROM olist_customer_dim AS cd
    JOIN olist_customers AS c
        ON c.customer_unique_id = cd.customer_unique_id
    LEFT JOIN order_totals AS ot
        ON ot.customer_id = c.customer_id
    GROUP BY c.customer_unique_id
),
segmented AS (
    SELECT
        cs.customer_unique_id,
        cs.n_orders,
        cs.monetary_value,
        CASE
            WHEN cs.last_order_date IS NULL THEN NULL
            ELSE (rd.as_of_date - cs.last_order_date)::INTEGER
        END AS recency_days,
        CASE
            WHEN cs.monetary_value >= 1000 THEN 'VIP'
            WHEN cs.last_order_date IS NOT NULL
                 AND (rd.as_of_date - cs.last_order_date)::INTEGER <= 90
                 AND cs.n_orders = 1 THEN 'new'
            WHEN cs.n_orders >= 2 OR cs.monetary_value >= 100 THEN 'regular'
            ELSE 'inactive'
        END AS segment_name,
        CASE
            WHEN cs.monetary_value >= 1000 THEN 5
            WHEN cs.monetary_value >= 500  THEN 4
            WHEN cs.n_orders >= 2          THEN 3
            WHEN cs.n_orders = 1           THEN 2
            ELSE 1
        END::SMALLINT AS rfm_score
    FROM customer_stats AS cs
    CROSS JOIN reference_date AS rd
)
INSERT INTO customer_segments (
    customer_unique_id,
    segment_name,
    rfm_score,
    recency_days,
    monetary_value,
    n_orders
)
SELECT
    customer_unique_id,
    segment_name,
    rfm_score,
    recency_days,
    monetary_value,
    n_orders
FROM segmented
ON CONFLICT (customer_unique_id) DO UPDATE
SET segment_name   = EXCLUDED.segment_name,
    rfm_score      = EXCLUDED.rfm_score,
    recency_days   = EXCLUDED.recency_days,
    monetary_value = EXCLUDED.monetary_value,
    n_orders       = EXCLUDED.n_orders,
    assigned_at    = NOW()
RETURNING
    customer_unique_id,
    segment_name,
    rfm_score,
    recency_days,
    monetary_value,
    n_orders
"""

with engine.begin() as conn:
    result = conn.exec_driver_sql(segmentation_sql)
    customer_segments_result = pd.DataFrame(result.fetchall(), columns=list(result.keys()))

print(len(customer_segments_result))
print(customer_segments_result['segment_name'].value_counts())
print(customer_segments_result['rfm_score'].value_counts().sort_index())
customer_segments_result.head()

96096
segment_name
regular     45982
inactive    40032
new          8891
VIP          1191
Name: count, dtype: int64
rfm_score
2    89064
3     2608
4     3233
5     1191
Name: count, dtype: int64


,customer_unique_id,segment_name,rfm_score,recency_days,monetary_value,n_orders
0,0000366f3b9a7992bf8c76cfdf3221e2,regular,2,161,141.90,1
1,0000b849f77a49e4a4ce2b2a4ca5be3f,inactive,2,164,27.19,1
2,0000f46a3911fa3c0805444483337064,inactive,2,587,86.22,1
3,0000f6ccb0745a6a4b88665a16c9f078,inactive,2,371,43.62,1
4,0004aac84e0df4da2b147fca70cf8255,regular,2,338,196.89,1


Наповнення `customer_segments`: CTE рахують суму замовлень, агрегують їх до `customer_unique_id` і обчислюють `recency_days`, `segment_name`, `rfm_score`; `RETURNING` повертає вставлені або оновлені рядки.

1. **Чому `customer_unique_id`.** `customer_id` створюється на кожне замовлення, тож сегментація за ним розбила б історію покупця, і `n_orders` майже завжди дорівнював би 1.
2. **Навіщо `olist_customer_dim` для FK.** FK може посилатися лише на `PRIMARY KEY`/`UNIQUE`, а в `olist_customers` `customer_unique_id` повторюється; `olist_customer_dim` має рівно один рядок на покупця.
3. **Чому `reference_date` з датасету, а не `CURRENT_DATE`.** Дані 2016–2018 років: від `CURRENT_DATE` recency становила б кілька років, а сегмент `new` був би порожнім. `MAX(order_purchase_timestamp)::DATE + 1` дає відтворюваний результат незалежно від дати запуску.
4. **Як `ON CONFLICT` робить сегментацію повторюваною.** Завдяки `UNIQUE (customer_unique_id)` повторний запуск не дублює рядки, а `DO UPDATE` перезаписує сегмент актуальними значеннями.

In [ ]:
pd.read_sql(text("""
SELECT
    conname,
    contype,
    pg_get_constraintdef(oid) AS definition
FROM pg_constraint
WHERE conrelid = 'customer_segments'::regclass
ORDER BY conname
"""), engine)

,conname,contype,definition
0,customer_segments_customer_unique_id_key,u,UNIQUE (customer_unique_id)
1,customer_segments_monetary_value_check,c,CHECK ((monetary_value >= (0)::numeric))
2,customer_segments_n_orders_check,c,CHECK ((n_orders >= 0))
3,customer_segments_pkey,p,PRIMARY KEY (segment_id)
4,customer_segments_recency_days_check,c,CHECK (((recency_days IS NULL) OR (recency_day...
5,customer_segments_rfm_score_check,c,CHECK (((rfm_score >= 1) AND (rfm_score <= 5)))
6,customer_segments_segment_name_check,c,"CHECK ((segment_name = ANY (ARRAY['VIP'::text,..."
7,fk_customer_segments_customer_unique,f,FOREIGN KEY (customer_unique_id) REFERENCES ol...


Перевірка через `pg_constraint`: видно identity PK, `UNIQUE (customer_unique_id)`, FK на `olist_customer_dim` з `ON DELETE CASCADE` і всі `CHECK`-обмеження (`contype`: `p` — PK, `u` — unique, `f` — FK, `c` — check).

# Завдання 4

In [ ]:
pd.read_sql(text("""
SELECT
    p.product_category_name,
    s.seller_state,
    COUNT(DISTINCT oi.order_id) AS n_orders,
    SUM(oi.price + oi.freight_value)::NUMERIC(12, 2) AS revenue
FROM olist_order_items AS oi
JOIN olist_orders AS o
    ON o.order_id = oi.order_id
JOIN olist_products AS p
    ON p.product_id = oi.product_id
JOIN olist_sellers AS s
    ON s.seller_id = oi.seller_id
WHERE o.order_status = 'delivered'
GROUP BY p.product_category_name, s.seller_state
ORDER BY revenue DESC NULLS LAST
LIMIT 10
"""), engine)

,product_category_name,seller_state,n_orders,revenue
0,cama_mesa_banho,SP,8214,1070048.14
1,relogios_presentes,SP,4494,1016168.80
2,beleza_saude,SP,5684,790967.66
3,esporte_lazer,SP,4772,680165.54
4,moveis_decoracao,SP,4597,605650.63
5,cool_stuff,SP,2621,484196.53
6,utilidades_domesticas,SP,4021,471810.66
7,automotivo,SP,3018,438833.22
8,informatica_acessorios,SP,2783,388506.42
9,bebes,SP,1981,339129.74


Які категорії товарів і штати продавців дають найбільший revenue у `delivered`-замовленнях?

`INNER JOIN` чотирьох таблиць; revenue = `price + freight_value`.

Усі топ-10 комбінацій — продавці зі SP. Лідер — `cama_mesa_banho` (1 070 048.14), далі `relogios_presentes` (1 016 168.80) при майже вдвічі меншій кількості замовлень, тобто з вищим середнім чеком.

In [ ]:
pd.read_sql(text("""
WITH order_totals AS (
    SELECT
        o.order_id,
        c.customer_unique_id,
        SUM(oi.price + oi.freight_value)::NUMERIC(12, 2) AS order_total
    FROM olist_orders AS o
    JOIN olist_customers AS c
        ON c.customer_id = o.customer_id
    LEFT JOIN olist_order_items AS oi
        ON oi.order_id = o.order_id
    GROUP BY
        o.order_id,
        c.customer_unique_id
),
review_per_order AS (
    SELECT
        rv.order_id,
        AVG(rv.review_score)::NUMERIC(3, 2) AS order_review_score
    FROM olist_order_reviews AS rv
    WHERE rv.review_score IS NOT NULL
    GROUP BY rv.order_id
),
customer_stats AS (
    SELECT
        ot.customer_unique_id,
        COUNT(DISTINCT ot.order_id) AS n_orders,
        COALESCE(SUM(ot.order_total), 0)::NUMERIC(12, 2) AS total_spend,
        AVG(rpo.order_review_score)::NUMERIC(3, 2) AS avg_review_score
    FROM order_totals AS ot
    LEFT JOIN review_per_order AS rpo
        ON rpo.order_id = ot.order_id
    GROUP BY ot.customer_unique_id
)
SELECT
    cd.customer_unique_id,
    cd.customer_state,
    COALESCE(cs.n_orders, 0) AS n_orders,
    COALESCE(cs.total_spend, 0)::NUMERIC(12, 2) AS total_spend,
    COALESCE(cs.avg_review_score, 0)::NUMERIC(3, 2) AS avg_review_score
FROM olist_customer_dim AS cd
LEFT JOIN customer_stats AS cs
    ON cs.customer_unique_id = cd.customer_unique_id
ORDER BY total_spend DESC NULLS LAST
LIMIT 20
"""), engine)

,customer_unique_id,customer_state,n_orders,total_spend,avg_review_score
0,0a0a92112bd4c708ca5fde585afaa872,RJ,1,13664.08,1.0
1,da122df9eeddfedc1dc1f5349a1a690c,RJ,2,7571.63,5.0
2,763c8b1c9c68a0229c42c9fc6f662b93,ES,1,7274.88,1.0
3,dc4802a71eae9be1dd28f5d788ceb526,MS,1,6929.31,5.0
4,459bef486812aa25204be022145caa62,ES,1,6922.21,0.0
5,ff4159b92c40ebe40454e3e6a7c35ed6,SP,1,6726.66,5.0
6,4007669dec559734d6f53e029e360987,MG,1,6081.54,1.0
7,5d0a2980b292d049061542014e8960bf,GO,1,4809.44,1.0
8,eebb5dda148d3893cdaf5b5ca3040ccb,SP,1,4764.34,4.0
9,48e1ac109decbb87765a3eade6854098,PB,1,4681.78,5.0


Які customer-level features (кількість замовлень, витрати, середня оцінка) можна побудувати на рівні `customer_unique_id`?

`LEFT JOIN` від `olist_customer_dim` зберігає всіх покупців, `COALESCE` заповнює пропуски нулями. `review_per_order` захищає від множення review score через позиції замовлення — та сама проблема грануляції, що й у Завданні 2.2.

Найбільший покупець (`0a0a9211…`) витратив 13 664.08 за одне замовлення й поставив оцінку 1.0. У топ-20 майже всі мають одне замовлення, оцінки поляризовані (5.0 або 1.0). `avg_review_score = 0.0` — результат `COALESCE` для покупця без відгуку, а не реальна оцінка.

In [ ]:
pd.read_sql(text("""
WITH customer_states AS (
    SELECT customer_state AS state, COUNT(*) AS n_customers
    FROM olist_customers
    GROUP BY customer_state
), seller_states AS (
    SELECT seller_state AS state, COUNT(*) AS n_sellers
    FROM olist_sellers
    GROUP BY seller_state
)
SELECT
    COALESCE(c.state, s.state) AS state,
    c.n_customers,
    s.n_sellers,
    CASE
        WHEN c.state IS NOT NULL AND s.state IS NOT NULL THEN 'both'
        WHEN c.state IS NOT NULL THEN 'customers_only'
        ELSE 'sellers_only'
    END AS state_presence
FROM customer_states AS c
FULL OUTER JOIN seller_states AS s
    ON s.state = c.state
ORDER BY state
"""), engine)

,state,n_customers,n_sellers,state_presence
0,AC,81,1.0,both
1,AL,413,NaN,customers_only
2,AM,148,1.0,both
3,AP,68,NaN,customers_only
4,BA,3380,19.0,both
5,CE,1336,13.0,both
6,DF,2140,30.0,both
7,ES,2033,23.0,both
8,GO,2020,40.0,both
9,MA,747,1.0,both


Які штати представлені тільки серед клієнтів, тільки серед продавців або в обох групах?

`FULL OUTER JOIN` агрегатів по штату зберігає рядки з обох боків; `CASE` класифікує присутність.

23 штати — `both`, 4 — `customers_only` (AL, TO, AP, RR), `sellers_only` немає. Тому нижче той самий `FULL OUTER JOIN` виконано на рівні міст.

In [ ]:
pd.read_sql(text("""
WITH customer_cities AS (
    SELECT customer_state AS state, customer_city AS city, COUNT(*) AS n_customers
    FROM olist_customers
    GROUP BY customer_state, customer_city
), seller_cities AS (
    SELECT seller_state AS state, seller_city AS city, COUNT(*) AS n_sellers
    FROM olist_sellers
    GROUP BY seller_state, seller_city
), city_presence AS (
    SELECT
        COALESCE(c.state, s.state) AS state,
        COALESCE(c.city, s.city) AS city,
        c.n_customers,
        s.n_sellers,
        CASE
            WHEN c.city IS NOT NULL AND s.city IS NOT NULL THEN 'both'
            WHEN c.city IS NOT NULL THEN 'customers_only'
            ELSE 'sellers_only'
        END AS city_presence
    FROM customer_cities AS c
    FULL OUTER JOIN seller_cities AS s
        ON s.state = c.state
       AND s.city = c.city
)
SELECT
    city_presence,
    COUNT(*) AS n_cities,
    COALESCE(SUM(n_customers), 0) AS n_customers,
    COALESCE(SUM(n_sellers), 0) AS n_sellers,
    (ARRAY_AGG(state || ' / ' || city ORDER BY COALESCE(n_sellers, 0) DESC, COALESCE(n_customers, 0) DESC, state, city))[1:5] AS top_examples
FROM city_presence
GROUP BY city_presence
ORDER BY city_presence
"""), engine)

,city_presence,n_cities,n_customers,n_sellers,top_examples
0,both,531,77269.0,2971.0,"[SP / sao paulo, PR / curitiba, RJ / rio de ja..."
1,customers_only,3779,22172.0,0.0,"[PA / belem, AL / maceio, RJ / rio das ostras,..."
2,sellers_only,105,0.0,124.0,"[SP / mogi guacu, SP / sp, SP / curitiba, SP /..."


Які міста мають тільки клієнтів, тільки продавців або обидві групи?

Ключ з'єднання — `(state, city)`.

531 місто в обох групах, 3 779 — лише з клієнтами, 105 — лише з продавцями. `sellers_only` значною мірою — брудні `seller_city` (`SP / sp`, `SP / sao paulo - sp`), які потребують нормалізації.

In [ ]:
pd.read_sql(text("""
SELECT
    s1.seller_id AS seller_a,
    s2.seller_id AS seller_b,
    s1.seller_state,
    s1.seller_city AS city_a,
    s2.seller_city AS city_b
FROM olist_sellers AS s1
JOIN olist_sellers AS s2
    ON s1.seller_state = s2.seller_state
   AND s1.seller_id < s2.seller_id
   AND s1.seller_city IS DISTINCT FROM s2.seller_city
ORDER BY s1.seller_state, seller_a, seller_b
LIMIT 20
"""), engine)

,seller_a,seller_b,seller_state,city_a,city_b
0,1444c08e64d55fb3c25f0f09c07ffcf2,2b402d5dc42554061f8ea98d1916f148,BA,lauro de freitas,irece
1,1444c08e64d55fb3c25f0f09c07ffcf2,398cb257329ef7af7f1943a8974a3cbc,BA,lauro de freitas,salvador
2,1444c08e64d55fb3c25f0f09c07ffcf2,4221a7df464f1fe2955934e30ff3a5a1,BA,lauro de freitas,bahia
3,1444c08e64d55fb3c25f0f09c07ffcf2,43753b27d77860f1654aa72e251a7878,BA,lauro de freitas,barro alto
4,1444c08e64d55fb3c25f0f09c07ffcf2,4aba391bc3b88717ce08eb11e44937b2,BA,lauro de freitas,arraial d'ajuda (porto seguro)
5,1444c08e64d55fb3c25f0f09c07ffcf2,4fb41dff7c50136976d1a5cf004a42e2,BA,lauro de freitas,feira de santana
6,1444c08e64d55fb3c25f0f09c07ffcf2,651530bf5c607240ccdd89a30c9c9712,BA,lauro de freitas,ipira
7,1444c08e64d55fb3c25f0f09c07ffcf2,659e8466eb3ff1b0e8740d74fb7bbedd,BA,lauro de freitas,eunapolis
8,1444c08e64d55fb3c25f0f09c07ffcf2,75d34ebb1bd0bd7dde40dd507b8169c3,BA,lauro de freitas,salvador
9,1444c08e64d55fb3c25f0f09c07ffcf2,a3dd39f583bc80bd8c5901c95878921e,BA,lauro de freitas,salvador


Які пари продавців з одного штату, але різних міст, можна порівнювати регіонально?

`s1.seller_id < s2.seller_id` прибирає дзеркальні пари, `IS DISTINCT FROM` залишає різні міста.

Перші 20 пар — штат BA; продавець `1444c08e…` з Lauro de Freitas утворює пари з продавцями Salvador, Feira de Santana тощо. «Місто» `bahia` показує, що брудні `seller_city` створюють хибні пари.

In [ ]:
pd.read_sql(text("""
SELECT
    o.order_id,
    o.customer_id,
    o.order_status,
    o.order_purchase_timestamp,
    COUNT(*) OVER () AS n_orders_without_review
FROM olist_orders AS o
LEFT JOIN olist_order_reviews AS rv
    ON rv.order_id = o.order_id
WHERE rv.review_row_id IS NULL
ORDER BY o.order_purchase_timestamp
LIMIT 20
"""), engine)

,order_id,customer_id,order_status,order_purchase_timestamp,n_orders_without_review
0,cda873529ca7ab71f677d5ec11a40304,76c74aaff2f3f7355f46d9818ad092b8,shipped,2016-10-05 16:57:30,768
1,5da2004d85f2828349d887c989da2adc,95f592086150f0c95d3082aea047b816,delivered,2016-10-05 20:20:59,768
2,02190241f7190a1f3c7e0df95a749c6a,55c9044401ffab0c717f24405e2e042f,delivered,2016-10-07 10:52:22,768
3,0efd0bc268d34da3f01f4ff25a6d4335,95446917717bb58d553d107d0f1668f6,shipped,2016-10-07 15:53:31,768
4,c776863a93dc0740c6e7d78104b21413,1710b798ebdc5db6665c815b90a64999,delivered,2016-10-10 15:01:27,768
5,d730d8d0ea3149120703933fdf98b2d4,ce2a66464da8b80dedd7639a6e489a58,shipped,2017-01-13 20:12:41,768
6,05cd156f0f47579a36d7a52724fc31b8,976224a95faf328b17654b1b1411d5e1,delivered,2017-01-14 18:27:28,768
7,719b940ab2f1e351ffb14a0a02ffe3c4,ede9a1daf58d256c0d648ef7e1479010,delivered,2017-01-16 13:01:51,768
8,fbd82b1fe98967de0db29b6cf134d00a,45f5c9d984dbfa2105e3544cc85133bf,delivered,2017-01-18 08:55:36,768
9,171398a0cd843c65ac78aee9e05a1f49,f0244902c07b7c0799e468218579a925,delivered,2017-01-18 16:07:37,768


Які замовлення не мають жодного review-запису?

Anti-join: `LEFT JOIN` + `rv.review_row_id IS NULL`; `COUNT(*) OVER ()` дає загальну кількість.

Без відгуку 768 замовлень із 99 441 (≈0.8%), більшість — `delivered`. Відсутню оцінку треба обробляти явно (як у 4.2).

# Завдання 5

In [ ]:
pd.read_sql(text("""
SELECT
    p.product_category_name,
    COUNT(DISTINCT oi.order_id) AS n_orders,
    SUM(oi.price)::NUMERIC(12, 2) AS gross_revenue,
    SUM(oi.freight_value)::NUMERIC(12, 2) AS freight,
    AVG(oi.price)::NUMERIC(8, 2) AS avg_item_price
FROM olist_order_items AS oi
JOIN olist_products AS p
    ON p.product_id = oi.product_id
GROUP BY p.product_category_name
HAVING SUM(oi.price) > 10000
ORDER BY gross_revenue DESC NULLS LAST
"""), engine)

,product_category_name,n_orders,gross_revenue,freight,avg_item_price
0,beleza_saude,8836,1258681.34,182566.73,130.16
1,relogios_presentes,5624,1205005.68,100535.93,201.14
2,cama_mesa_banho,9417,1036988.68,204693.04,93.30
3,esporte_lazer,7720,988048.97,168607.51,114.34
4,informatica_acessorios,6689,911954.32,147318.08,116.51
5,moveis_decoracao,6449,729762.49,172749.30,87.56
6,cool_stuff,3632,635290.85,84039.10,167.36
7,utilidades_domesticas,5884,632248.66,146149.11,90.79
8,automotivo,3897,592720.11,92664.21,139.96
9,ferramentas_jardim,3518,485256.46,98962.75,111.63


Які категорії дають gross revenue понад 10 000 і яка в них кількість замовлень, доставка та середня ціна?

Поріг проходять 53 групи з 74. Лідер — `beleza_saude` (1 258 681.34), далі `relogios_presentes` (1 205 005.68) з високою середньою ціною позиції 201.14. Рядок `NaN` — товари без категорії (179 535.28).

**`WHERE` vs `HAVING`:** `WHERE` фільтрує рядки **до** агрегації, `HAVING` — групи **після**. `HAVING SUM(oi.price) > 10000` перевіряє суму категорії, а `WHERE oi.price > 10000` перевіряв би ціну окремої позиції — зовсім інший результат.

In [ ]:
pd.read_sql(text("""
WITH order_totals AS (
    SELECT
        o.order_id,
        c.customer_state,
        SUM(oi.price + oi.freight_value)::NUMERIC(12, 2) AS order_total
    FROM olist_orders AS o
    JOIN olist_customers AS c
        ON c.customer_id = o.customer_id
    JOIN olist_order_items AS oi
        ON oi.order_id = o.order_id
    GROUP BY
        o.order_id,
        c.customer_state
)
SELECT
    customer_state,
    COUNT(*) AS n_orders,
    SUM(order_total)::NUMERIC(12, 2) AS gmv,
    (SUM(order_total) / NULLIF(COUNT(*), 0))::NUMERIC(8, 2) AS aov
FROM order_totals
GROUP BY customer_state
HAVING COUNT(*) >= 100
ORDER BY aov DESC NULLS LAST
"""), engine)

,customer_state,n_orders,gmv,aov
0,PB,532,140987.81,265.01
1,AL,411,96229.40,234.13
2,RO,247,57558.02,233.03
3,PA,970,217647.11,224.38
4,TO,279,61354.42,219.91
5,PI,493,108132.28,219.34
6,SE,345,73032.32,211.69
7,RN,482,101895.08,211.40
8,CE,1327,275606.30,207.69
9,MT,903,186168.96,206.17


Який середній чек (AOV) у кожному штаті покупців і які штати дають найбільший GMV?

CTE `order_totals` спершу зводить позиції до замовлення; `HAVING COUNT(*) >= 100` прибирає штати з малою вибіркою.

Відсіяно RR, AP і AC. Найвищий AOV — у віддалених штатах (PB 265.01, AL 234.13), найнижчий — у SP (143.12), хоча SP дає найбільший GMV (5 921 678.12).

In [ ]:
pd.read_sql(text("""
SELECT
    c.customer_unique_id,
    COUNT(DISTINCT o.order_id) AS n_orders,
    SUM(oi.price)::NUMERIC(12, 2) AS total_spend,
    STRING_AGG(
        DISTINCT COALESCE(p.product_category_name, '(unknown)'),
        ', ' ORDER BY COALESCE(p.product_category_name, '(unknown)')
    ) AS categories
FROM olist_customers AS c
JOIN olist_orders AS o
    ON o.customer_id = c.customer_id
JOIN olist_order_items AS oi
    ON oi.order_id = o.order_id
LEFT JOIN olist_products AS p
    ON p.product_id = oi.product_id
GROUP BY c.customer_unique_id
HAVING COUNT(DISTINCT o.order_id) >= 1
ORDER BY total_spend DESC NULLS LAST
LIMIT 20
"""), engine)

,customer_unique_id,n_orders,total_spend,categories
0,0a0a92112bd4c708ca5fde585afaa872,1,13440.00,telefonia_fixa
1,da122df9eeddfedc1dc1f5349a1a690c,2,7388.00,eletroportateis
2,763c8b1c9c68a0229c42c9fc6f662b93,1,7160.00,telefonia_fixa
3,dc4802a71eae9be1dd28f5d788ceb526,1,6735.00,utilidades_domesticas
4,459bef486812aa25204be022145caa62,1,6729.00,pcs
5,ff4159b92c40ebe40454e3e6a7c35ed6,1,6499.00,artes
6,4007669dec559734d6f53e029e360987,1,5934.60,agro_industria_e_comercio
7,eebb5dda148d3893cdaf5b5ca3040ccb,1,4690.00,eletroportateis
8,5d0a2980b292d049061542014e8960bf,1,4599.90,cool_stuff
9,48e1ac109decbb87765a3eade6854098,1,4590.00,pcs


Які категорії купував кожен реальний покупець (`customer_unique_id`) і скільки він витратив?

`STRING_AGG(DISTINCT ...)` збирає унікальні категорії покупця в один рядок, `COALESCE` підставляє `(unknown)` для товарів без категорії.

У топ-20 кожен покупець купував лише одну категорію; лідер `0a0a9211…` — 13 440.00 на `telefonia_fixa`. Загалом лише 2 327 з 95 420 покупців (≈2.4%) купували більше однієї категорії.

# Завдання 6

In [ ]:
pd.read_sql(text("""
SELECT customer_state AS state, 'customer' AS role
FROM olist_customers

UNION

SELECT seller_state AS state, 'seller' AS role
FROM olist_sellers
ORDER BY state, role
"""), engine)

,state,role
0,AC,customer
1,AC,seller
2,AL,customer
3,AM,customer
4,AM,seller
5,AP,customer
6,BA,customer
7,BA,seller
8,CE,customer
9,CE,seller


У яких штатах платформа присутня через покупців, продавців або обидві сторони?

**Чому `UNION`:** потрібне об'єднання множин без повторів; `UNION ALL` повернув би 102 536 рядків з дублікатами. Колонка `role` зберігає джерело штату.

50 рядків — 27 штатів `customer` і 23 `seller`; AL, AP, RR, TO мають лише `customer`, що збігається з 4.3.

In [ ]:
pd.read_sql(text("""
SELECT customer_state AS state
FROM olist_customers

INTERSECT

SELECT seller_state AS state
FROM olist_sellers
ORDER BY state
"""), engine)

,state
0,AC
1,AM
2,BA
3,CE
4,DF
5,ES
6,GO
7,MA
8,MG
9,MS


У яких штатах є і покупці, і продавці?

**Чому `INTERSECT`:** потрібен саме перетин множин; `INTERSECT` сам повертає унікальні значення без `JOIN` і `DISTINCT`.

23 штати — стільки ж, скільки `both` у 4.3.

In [ ]:
pd.read_sql(text("""
WITH ranked_orders AS (
    SELECT
        c.customer_unique_id,
        o.customer_id,
        o.order_id,
        o.order_purchase_timestamp,
        ROW_NUMBER() OVER (
            PARTITION BY c.customer_unique_id
            ORDER BY o.order_purchase_timestamp DESC, o.order_id DESC
        ) AS rn,
        COUNT(*) OVER (PARTITION BY c.customer_unique_id) AS n_orders
    FROM olist_orders AS o
    JOIN olist_customers AS c
        ON c.customer_id = o.customer_id
)
SELECT
    customer_unique_id,
    customer_id,
    order_id,
    order_purchase_timestamp,
    n_orders,
    COUNT(*) OVER () AS n_customers
FROM ranked_orders
WHERE rn = 1
ORDER BY
    n_orders DESC,
    customer_unique_id
LIMIT 20
"""), engine)

,customer_unique_id,customer_id,order_id,order_purchase_timestamp,n_orders,n_customers
0,8d50f5eadf50201ccdcedfb9e2ac8455,a682769c4bc10fc6ef2101337a6c83c9,d3582fd5ccccd9cb229a63dfb417c86f,2018-08-20 19:14:26,17,96096
1,3e43e6105506432c953e165fb2acf44c,1815fdaaae692479110428598596b5e5,ff89ef7b3952bba5ac06d61c4a79ffbe,2018-02-27 18:36:39,9,96096
2,1b6c7548a2a1f9037c1fd3ddfed95f33,fff7466a253c0e59499ea943462c10f9,43f08913407cac4e74a898d968e58c1a,2018-02-14 13:22:12,7,96096
3,6469f99c1f9dfae7733b25662e7f1782,4478700ff008d3fffa02e95d89b9795f,ccb1046449f1a494de800567e7cb5979,2018-06-28 00:43:34,7,96096
4,ca77025e7201e3b30c44b472ff346268,6ccedfba5919d72fcc8c51bfa982de62,8dacd1da7c4dc40d5c59e7ddd93b3ac8,2018-06-01 11:38:29,7,96096
5,12f5d6e1cbf93dafd9dcc19095df0b3d,ed1793d2d1e4175d5846ce7ebb4a01f5,747996a66f5aa711deb8ae58f5ae46a0,2017-01-05 15:25:10,6,96096
6,47c1a3033b8b77b3ab6e109eb4d5fdf3,5daf8c734e89f88e5305f9283e7ca135,b86ec88f5bffbf1f61fb5daed332e355,2018-01-24 15:15:26,6,96096
7,63cfc61cee11cbe306bff5857d00bfe4,906614004e38a1e922ff033563bfdcf2,c29713133e5e33937af1a6e1e1370552,2018-05-28 17:20:02,6,96096
8,dc813062e0fc23409cd255f7f53c7074,4fd0b9bb860c9a3d841746e1cb742314,f0031384d36ff045ee1e866942d0fa7a,2018-08-23 00:07:26,6,96096
9,de34b16117594161a6a89c50b289d35a,cbbd508281ea66e2f36b4acfc28a85b0,d29eca3633b0b60adf268cae9133cd93,2018-01-17 23:34:21,6,96096


Яке найновіше замовлення кожного реального покупця (`customer_unique_id`) і скільки всього замовлень він зробив?

`ROW_NUMBER()` нумерує замовлення покупця від найновішого; `order_id DESC` — детермінований tie-break. `COUNT(*) OVER (PARTITION BY ...)` додає `n_orders`.

Після `rn = 1` залишається 96 096 рядків — кількість унікальних покупців. Найактивніший `8d50f5ea…` зробив 17 замовлень, останнє — 2018-08-20.

# Summary

1. Найскладнішим був full outer join: на рівні штатів він не дав жодного рядка тільки з продавцями, тож довелося повторити той самий запит на рівні міста, щоб побачити всі три групи результату.

2. Найважче було сформулювати бізнес-питання для self join продавців: пари продавців з одного штату, але різних міст, мають сенс лише як пошук регіональних логістичних партнерів, а брудні назви міст створювали хибні пари.

3. Найкориснішою для production ETL здається on conflict: upsert-запити зробили клітинки ідемпотентними, а returning одразу показував, що саме змінилося.

4. Window-функції можна використати для порядкового номера і кількості попередніх замовлень покупця, інтервалів між покупками, ковзної середньої оцінки продавця та фіч на певну дату без витоку даних з майбутнього.

5. Серед проблем якості даних: ідентифікатор відгуку не унікальний, 98410 значень на 99224 рядки, 768 замовлень без жодного відгуку, кілька продавців без жодної оцінки, товари без категорії, брудні назви міст продавців і помилка в назві колонки довжини товару.

6. Для фіч на рівні покупця використано унікальний ідентифікатор людини, а не ідентифікатор замовлення, бо кожне замовлення створює свій ідентифікатор клієнта: 99441 такий ідентифікатор відповідає лише 96096 реальним покупцям. Рахуючи фічі за ідентифікатором замовлення, історію покупця розбили б на частини, а кількість замовлень і суму витрат занизили б.

7. Перед обчисленням оцінки продавця дані потрібно було звести до рівня продавець-замовлення, бо таблиця товарів у замовленні має рядок на кожну позицію, а відгук належить замовленню в цілому. Без усунення дублів оцінка замовлення з кількома товарами врахувалася б кілька разів і зсунула б середню оцінку продавця.